presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -321,6 +341,28 @@ public void fail()
         }
     }
 
+    @Override
+    public void setNoMorePagesForLifespan(Lifespan lifespan)
+    {
+        noMorePagesForLifespan.add(lifespan);
+    }
+
+    @Override
+    public void registerLifespanFinishCallback(Consumer<Lifespan> callback)
+    {
+        checkState(lifespanFinishedCallback == null);
+        this.lifespanFinishedCallback = callback;
+    }
+
+    @Override
+    public boolean isLifespanFinished(Lifespan lifespan)
+    {
+        if (!noMorePagesForLifespan.contains(lifespan)) {
```

Comment:
Isn't this just 

```
return noMorePagesForLifespan.contains(lifespan) &&
        outstandingPagePerLifespan.get(lifespan).get() == 0;
```

Reply:
I noted a lifespan might never has any page enqueue.  In that case `outstandingPageCountPerLifespan.get(lifespan)` will return `null`. We should use some way similar to `PartitionedOutputBuffer`. Sorry for the confusion.